src/Uri.php

```diff
@@ -594,6 +598,25 @@ private function filterPath(string $path) : string
         return '/' . ltrim($path, '/');
     }
 
+    /**
+     * Encode invalid UTF-8 characters in given string. All other characters are unchanged.
+     */
+    private function filterInvalidUtf8(string $string) : string
+    {
+        if (preg_match('//u', $string)) {
+            return $string;
+        }
+
+        $letters = str_split($string);
```

Comment:
Noting here, because I had to look it up - `str_split()` splits into bytes instead of characters when it encounters a multibyte string.